# Day 4: Camp Cleanup
## Part One


In [1]:
## In how many assignment pairs does one range fully contain the other? ##

In [45]:
# Global variables

## Hold number of contained pairs
contained_pairs = 0

## Read input.txt and strip '\n'
with open('04 input.txt', 'r') as f:
    input = f.readlines()
input = map(str.strip, input)

In [46]:
# Functions

## Return int: index of dash char in elf assignment
def dash_finder(s):
    return s.find('-')

## Return tuple: lower and upper bound of assignment
def assignment_bounds(s):
    dash_index = dash_finder(s)
    lower_bound = int(s[:dash_index])
    upper_bound = int(s[dash_index+1:])
    return (lower_bound, upper_bound)

## Return bool: tuple1 within tuple2
def tuple_contained(t1, t2):
    lower = t1[0] >= t2[0]
    upper = t1[1] <= t2[1]
    return lower and upper

In [47]:
# Main

## Prepare input for processing
pair_list = map(lambda x : str.split(x, sep = ','), input)

## Iterate over elf_list
for elf1, elf2 in pair_list:
    elf1_ass = assignment_bounds(elf1) #ass for assignments, obviously
    elf2_ass = assignment_bounds(elf2)
    if tuple_contained(elf1_ass, elf2_ass) or tuple_contained(elf2_ass, elf1_ass):
        contained_pairs += 1
        
print(contained_pairs)


485


Correct!

## Part Two
In how many assignment pairs do the ranges overlap?

In [75]:
# Global variables

## Hold number of contained pairs
overlap_pairs = 0

## Read input.txt and strip '\n'
with open('04 input.txt', 'r') as f:
    input = f.readlines()
input = map(str.strip, input)

In [76]:
# Functions

## Return int: index of dash char in elf assignment
def dash_finder(s):
    return s.find('-')

## Return range: lower and upper bound of assignment
def assignment_range(s):
    dash_index = dash_finder(s)
    lower_bound = int(s[:dash_index])
    upper_bound = int(s[dash_index+1:])
    return range(lower_bound, upper_bound+1)

## Return bool: overlap between range1 and range2 larger than 0
def range_overlapping(r1, r2):
    overlap = range(max(r1.start, r2.start), min(r1.stop, r2.stop)) #ty stackoverflow
    return len(overlap) > 0


In [77]:
# Main

## Prepare input for processing
pair_list = map(lambda x : str.split(x, sep = ','), input)

## Iterate over elf_list
for elf1, elf2 in pair_list:
    elf1_range = assignment_range(elf1)
    elf2_range = assignment_range(elf2)
    if range_overlapping(elf1_range, elf2_range):
        overlap_pairs += 1
        #print(elf1, elf2) #for debugging purposes
        
print(overlap_pairs)

857


Made an initial mistake in the ranges: should have thought of the fact that the upper bound on a range is non-inclusive. 857 is correct.